Simple FoldCase for Turkish-I issue #19883

fbacher opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->

This change improves on patch for Turkish-I problem (19883). This fix is inferior to using Unicode ICUC4 library, but time does not now permit that change. (I started a PR for ICU4C #21690, but I think that I should delete it and create new one when appropriate, especially since numerous changes will be required to the PR).

The FoldCase here simply performs a ToLower using the locale en_GB.UTF8. The Turkish-I problem caused Kodi to be non-functional due to ToLower not treating ASCII i/I and Turkish "I" characters in a "normal" manner. This caused all sorts of problems when the key to lookup tables, etc. were normalized to lower case (or upper case). By specifying en_GB.UTF8 for Folding this problem is averted.

This will ONLY work with ASCII characters (or whatever en_GB defines). This should be sufficient, but there may be some odd cases where a key does not obey this rule. To address that problem, ICU should be used.

Many simple changes were required in code to convert all of the ToLower/ToUpper usages to FoldCase/FoldCaseUpper. Another change made was to use string_view instead of string for arguments and to return values by function return instead of by reference. This is in keeping with Kodi guidelines and performance improvement.

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Provides a better fix to #19883. It is still imperfect since it only address case folding and only works for characters in en_GB locale. But this should get the lion's share of problems and is definitely better than fix made in Matrix.
Problems not addressed by fix include: 
- non-ASCII characters (should be rare for key values, but not impossible)
- Does not address changes of case changing length of string
- General Unicode issues. C++ Unicode support is weak.
- 
## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Ran Kodi using various languages, including Turkish on Ubuntu 22.10 (be sure to install Turkish on Ubuntu and set LANG=tr_TR.
Added several FoldCase test cases to TestStringUtils. Did not add tests for wstring. Code is almost identical.  Ran all tests from make check.
(This code was based on my ICU based solution where I ran many tests over months.)

I did spot one POSSIBLE regression, but I don't think so. Currently Kodi nightly builds don't like to install on Ubuntu and blow up with a missing or broken wayland lib. My builds work just fine, EXCEPT when on the video top level menu item I don't see icons in the main window (Catagories: Movies, Files). I suspect that I don't have the right images installed.

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->

Allows Kodi to run in Turkish and several related languages.  

## Screenshots (if appropriate):
Image showing missing icons mentioned above. 
![image](https://user-images.githubusercontent.com/10661641/194437528-64ef4a2a-3854-4bd0-8056-08fb179b71bf.png)

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [x] I have added tests to cover my change
- [x] All new and existing tests passed

lrusak commented on xbmc/utils/StringUtils.cpp:
```diff
@@ -336,76 +176,100 @@ int compareWchar (const void* a, const void* b)
+  const std::ctype<char>& ct = use_facet<std::ctype<char> >(loc);
+  std::string output{input};;
+  transform(input.begin(), input.end(), output.begin(),
+      std::bind1st(std::mem_fun(&std::ctype<char>::tolower), &ct));
```
`std::bind1st` isn't available in c++17

ref: https://en.cppreference.com/w/cpp/utility/functional/bind12

lrusak replied:
`std::mem_fun` isn't available in c++17

https://en.cppreference.com/w/cpp/utility/functional/mem_fun

lrusak commented on xbmc/utils/StringUtils.cpp:
```diff
@@ -336,76 +176,100 @@ int compareWchar (const void* a, const void* b)
-    return *(unicode_lowers + (p - unicode_uppers));
-
-  return c;
+  const std::ctype<char>& ct = use_facet<std::ctype<char> >(loc);
```
`use_facet` should be in the `std` namespace, `std::use_facet`

ref: https://en.cppreference.com/w/cpp/locale/use_facet

fbacher replied:
Apparently I don't have my build config set to consider this as an error. Good catch. I have reworked this to not to be so fancy (I have not mastered C++ to be good at facets, etc.. Especially when I would like pointers to instances of functions which use a template to specify both locale and tolower, etc..)

lrusak reviewed (commented):
I'm not sure this PR is in a reviewable state. It makes some changes to very important code paths. If possible please break it up into smaller chunks with minimal changes per commit.

I'm also not sure that hardcoding the use of `std::locale("en_GB.UTF8")` is an acceptable solution.

CastagnaIT commented on xbmc/utils/StringUtils.h:
```diff
@@ -90,14 +91,129 @@ class StringUtils
+    * \return Case folded version of str (all lower-case)
+    */
+
+  static std::string FoldCase(std::string_view str);
```
i wondering if it is possible find a more appropriate name for these new methods
i find these names not immediately comprehensible

CastagnaIT commented:
here the `ToLower` is replaced by `FoldCase` in many cases (or all cases?)
are we sure it is required to be replaced everywhere?
for example
i have seen it replaced on some cases of directory/file source code, could cause side effects since file/dir names could change chars?

so i ask myself in which situations should a developer use ToLower or CaseFold?

fbacher commented:
I use FoldCase method name in keeping with ICUC4 (https://unicode-org.github.io/icu/userguide/transforms/casemappings.html). I invented FoldCaseUpper, which I think now was a mistake, since I have not seen an equivalent term used elsewhere. At least with ICU, UnicodeString::foldCase essentially produces lower case version of the input string (however, it is locale agnostic and there is more complexity to it than my simple ToLower('x', locale("en_GB.UTF8")). Even though the FoldCase here is only approximating a true fold case, I hope to replace it with ICU's foldCase in the future. If this name still seems inappropriate, I can suggest several other names. As for FoldCaseUpper, I'll have to give that some thought and a bit more code research. It is fairly rarely used in Kodi. I remember it being used when upper casing a file name (or extension). Presumably historically the filenames are always stored in upper case (unless it lower-cases them later). I'll have to dig more into this. I was just trying to preserve the existing behavior while also fixing the Turkish-I problem.

fbacher commented:
I will break up this commit into multiple commits (perhaps 10?) so that it can be reviewed in a more orderly and digestible fashion. I'm ashamed that this did not occur to me earlier. I was thinking that a single commit should be a complete, usable chunk of code, and that the defect could not be fixed by a single chunk. I then realized that the fix depends upon the entire PR and not a single commit. 

Each commit will be buildable, testable and runnable. It just won't fix the defect in isolation.

fbacher commented:
I should have explained the problem and this solution more clearly. Reading the issue is tedious and long.

The central problem is that in a few languages, like Turkic, ToLower and ToUpper behave in an unusual manner. In Turkic, ToUpper( Latin (ASCII) small letter ‘i’) becomes the capital letter ‘I’ with dot above ( \u0130 ‘İ’). Further, if memory serves me right, in Turkic ToLower(ToUpper('i')) != 'i'. 

This little problem broke all code such as: key = ToLower(value), where 'key' is used to do a lookup into a table or similar. Most uses of ToLower in Kodi is for such a purpose. The only time you need to use ToLower is for display to a user. If you are using it for lookup or comparison then you should use FoldCase, because lookup and comparison requires a caseless comparison of the values (except for the most unusual circumstances, which I did not notice). Of course it is possible that I missed something, but I believe that it would be a poor programming practice not to do caseless comparison.

The patch made in matrix was essentially to prevent the locale from being changed to any language where ToUpper('i') != 'I'. Further, the kodi startup script unset the LANG environment variable, causing the global locale to be 'C'. This broke the kodi python API getLanguage so that a plugin could not determine its locale which caused me much grief. Further, it caused Python's startup code to set the default locale for filesystem names to be ASCII, requiring every filesystem open operation to specifically specify the locale to override. I will submit another PR or two to address these problems.

I hope that this clarifies things a bit.

fbacher committed:
Stringutils new & modified APIs

- Created StringUtils::FoldCase to use instead of ToLower to 'normalize'
key ids.
- Modified ToLower/ToUpper to use string_view and to return by value
instead of by reference.

Modified a subset of code: utils & utils/test to use these APIs.

fbacher commented:
I made the following updates (I hope that I rebased things correctly to show only one commit)
I rolled back all of my changes and then:

- Added StringUtils::FoldCase and FoldCaseUpper (Yeah, ended up keeping it because it saves making a number of little changes to work around using "lower case" instead. 
- Modified StringUtils::ToLower & ToUpper so that each has a signature using string_view/wstring_view and also returns by value and not by reference. (Old signatures kept as necessary to prevent from breaking code)
- Changed xbmc/utils and xbmc/utils/test to use FoldCase

Hopefully this is a bite sized chunk. This took me a while to do because I researched implementing a little better FoldCase as well as the presence of two tables (char arrays) in StringUtils.cpp: unicode_uppers & unicode_lowers. I documented details in the code and it is hopefully clear enough. Here is a summary:

After looking at ICU4C Case folding it became clear to me that the existing unicode_uppers & unicode_lowers were designed for case folding. I replaced the now 10-year old tables with a pair made in 2021. FoldCase now uses these.

The existing code has the wstring versions of ToUpper/ToLower use unicode_uppers & unicode_lowers. This seems like a mistake since std:tolower/toupper is locale sensitive. If nothing else this means that you can get inconsistent results depending upon if the UTF8 or wstring version of ToUpper/ToLower is used. Baring objection I will make ToUpper/ToLower follow the C++ standard.

In a related topic, I have ToUpper, ToLower and FoldCase convert the strings passed to them to u32string to avoid the UTF8 and wstring versions of these methods returning different results for the same input due to C++ limitation of operating on a single code-unit (8-bits for UTF8) at a time. I performed a test using Russian and there were a number of discrepancies between the UTF8 and wstring versions of these methods. It is more expensive to do these conversions, but I don't think that it is a large issue. I'll examine it more before completing the PR. I added suggestions in the code.

If this commit goes well, I'll proceed with the others. For the most part they will just be batches of monotonous changes to use FoldCase.

fbacher commented:
Messed up squash in previous commit. Should now have only one commit.

fbacher commented:
Although I have been quiet, I have been very busy working on this . My assumption is that  any changes that I make at this point won't go into the product until after release and very likely not until the next release (please correct me if I am wrong).

Here is what I think the process is:

- I commit a complete logical unit of code that is bite-sized. For example, the first drop would be for implementing the FoldCase API, without any code using 
-  The next commit would be for adding tests for FoldCase
- The next commit would be for changing a dozen calls to ToLower to use FoldCase

Would each commit be merged into master after approval, or would the entire PR be merged when complete? Or something in between? I ask because not merging  with master would create a real headache over time merging master into the PR code simply because there will be lots of little changes throughout the code. 

Also, since I will be replacing all of the code that I have put in this PR, should I delete the PR and start over, or rebase, etc. the existing branch to start over?

Thanks for your patience.

fuzzard commented:
You are correct, this change wont make V20. The nature of very large invasive changes like this mean they will always need very extensive testing.

The key thing to get a huge change in is to try and make it as reviewable as possible. If its one single commit with 10k+ lines changed, its very difficult to thoroughly review, and takes an immense amount of time for reviewers. Its just less likely to get done in a timely fashion.

Think about how to break things down into bite sized chunks. Ideally every commit should not leave the branch in an unbuildable/testable state (in the event of a git bisect for tracking down issues)

A large change can be done over several PR's if theres a clean and clear approach taken. One thing to be aware of, is we try to avoid code being merged that isnt actively used. That is, a PR that introduces a function, without it being used, is less likely to be merged. This goes for both team members and other devs, but doubly so for other devs. Everyone provides their time as, and when, they can, so merging code that may never be used because a dev loses interest/other life commitments occur but promised "it'll come" at the start obviously doesnt work well.

Your choice if you reuse this PR, but PR's are free. theres no hassle in closing this out and opening another one if you feel you have a better approach than the current PR.

fbacher commented:
Abandoning this doe to major rework of packaging of commits and improved FoldCase algorithm/data
